integreat_cms/deepl_api/utils.py

```diff
@@ -0,0 +1,161 @@
+        """
+        self.translator = deepl.Translator(settings.DEEPL_AUTH_KEY)
+
+    def check_availability(self, request, language_slug):
+        """
+        This function checks, if the selected language is supported by DeepL
+
+        :param request: request that was sent to the server
+        :type request: ~django.http.HttpRequest
+
+        :param language_slug: current language slug
+        :type language_slug: str
+
+        :return: true or false
+        :rtype: bool
+        """
+        supported_source_languages = [
+            source_language.code.lower()
+            for source_language in self.translator.get_source_languages()
+        ]
+        supported_target_languages = [
+            target_languages.code.lower()[:2]
+            for target_languages in self.translator.get_target_languages()
+        ]
+        return (
+            request.region.get_source_language.slug in supported_source_languages
+            and language_slug in supported_target_languages
+        )
+
+    def get_translation(self, content, source_language_slug, target_language_key):
+        """
+        This functions gets the translation from DeepL
+
+        :param content: passed content
+        :type content: str
+
+        :param source_language_slug: slug of source language
+        :type source_language_slug: str
+
+        :param target_language_key: slug or in case of pt and en bcp of target language
+        :type target_language_key: str
+
+        :return: translated post in target language
+        :rtype: str
+        """
+        translated_content = self.translator.translate_text(
+            content, source_lang=source_language_slug, target_lang=target_language_key
+        )
+        return translated_content
+
+    def deepl_translation(self, request, content_objects, language_slug, form_class):
+        """
+        This functions gets the translation from DeepL
+
+        :param request: passed request
+        :type request: ~django.http.HttpRequest
+
+        :param content_objects: passed content objects
+        :type content_objects: ~django.db.models.query.QuerySet [ ~integreat_cms.cms.models.abstract_content_model.AbstractContentModel ]
+
+        :param language_slug: current GUI language slug
+        :type language_slug: str
+
+        :param form_class: passed Form class of content type
+        :type form_class: ~integreat_cms.cms.forms.custom_content_model_form.CustomContentModelForm
+        """
+        # Get target language
+        target_language = request.region.get_language_or_404(language_slug)
+        source_language = request.region.get_source_language(language_slug)
+        for content_object in content_objects:
+            if source_translation:
+                source_translation = content_object.get_translation(
+                    source_language.slug
+                )
+            else:
+                messages.error(request, _("No source translation could be found."))
```

Comment:
```suggestion
            source_translation = content_object.get_translation(source_language.slug)
            if not source_translation:
                messages.error(
                    request,
                    _('No source translation could be found for {} "{}".').format(
                        type(content_object)._meta.verbose_name.title(),
                        content_object.best_translation.title,
                    ),
                )
                continue
```
You need to reverse the order here, first you need to get the translation, then you can ceck whether the translation exists. Otherwise you get the error `UnboundLocalError: local variable 'source_translation' referenced before assignment`. Also, you need to do a `continue` to skip the rest of the loop-iteration in case of an error, otherwise it will add the error message, but will just continue with the code until the `AttributeError` occurs.
Apart from that, I'd also add the "best translation title" to the message in case multiple objects were selected and it's not clear which of them were successful and which of them failed.